job.go

```diff
@@ -0,0 +1,253 @@
+	Output           struct {
+		Error  string `bson:"error" json:"error" yaml:"error"`
+		Output string `bson:"output" json:"output" yaml:"output"`
+	} `bson:"output" json:"output" yaml:"output"`
+	ExitCode      int `bson:"exit_code" json:"exit_code" yaml:"exit_code"`
+	OutputOptions struct {
+		SuppressOutput    bool `bson:"suppress_output,omitempty" json:"suppress_output,omitempty" yaml:"suppress_output,omitempty"`
+		SuppressError     bool `bson:"suppress_error,omitempty" json:"suppress_error,omitempty" yaml:"suppress_error,omitempty"`
+		SendOutputToError bool `bson:"redirect_output_to_error,omitempty" json:"redirect_output_to_error,omitempty" yaml:"redirect_output_to_error,omitempty"`
+		SendErrorToOutput bool `bson:"redirect_error_to_output,omitempty" json:"redirect_error_to_output,omitempty" yaml:"redirect_error_to_output,omitempty"`
+	} `bson:"output_opts,omitempty" json:"output_opts,omitempty" yaml:"output_opts,omitempty"`
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	makep ProcessConstructor
+}
+
+const (
+	amboyJobName                     = "jasper-shell-job"
+	amboySimpleCapturedOutputJobName = "jasper-simple-shell-job"
+	amboyForegroundOutputJobName     = "jasper-foreground-job"
+)
+
+func RegisterJobs(pc ProcessConstructor) {
+	registry.AddJobType(amboyJobName, func() amboy.Job { return amboyJobFactory(pc) })
+	registry.AddJobType(amboySimpleCapturedOutputJobName, func() amboy.Job { return amboySimpleCapturedOutputJobFactory(pc) })
+	registry.AddJobType(amboyForegroundOutputJobName, func() amboy.Job { return amboyForegroundOutputJobFactory(pc) })
+}
+
+func amboyJobFactory(pc ProcessConstructor) *amboyJob {
+	j := &amboyJob{
+		makep:    pc,
+		ExitCode: -1,
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    amboyJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+func NewJob(pc ProcessConstructor, cmd string) amboy.Job {
+	j := amboyJobFactory(pc)
+	j.CmdString = cmd
+	j.SetID(fmt.Sprintf("%s.%x", amboyJobName, sha1.Sum([]byte(cmd))))
+	return j
+}
+
+func NewJobBasic(cmd string) amboy.Job {
+	j := amboyJobFactory(newBasicProcess)
+	j.CmdString = cmd
+	j.SetID(fmt.Sprintf("%s.basic.%x", amboyJobName, sha1.Sum([]byte(cmd))))
+	return j
+}
+
+func NewJobExtended(pc ProcessConstructor, cmd string, env map[string]string, wd string) amboy.Job {
+	j := amboyJobFactory(pc)
+	j.CmdString = cmd
+	j.Environment = env
+	j.WorkingDirectory = wd
+	j.SetID(fmt.Sprintf("%s.ext.%x", amboyJobName, sha1.Sum([]byte(cmd))))
+	return j
+}
+
+func NewJobBasicExtended(cmd string, env map[string]string, wd string) amboy.Job {
+	j := amboyJobFactory(newBasicProcess)
+	j.CmdString = cmd
+	j.Environment = env
+	j.WorkingDirectory = wd
+	j.SetID(fmt.Sprintf("%s.ext.%x", amboyJobName, sha1.Sum([]byte(cmd))))
+	return j
+}
+
+func (j *amboyJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.ExitCode >= 0 {
```

Comment:
The exit code returned can be -1 if the `Wait` below this has its context cancelled, even though the process has been created. Is this okay?

Reply:
There's probably no great way to handle that case?